### PR TITLE
Allow Operator Generated bootstrap token

### PR DESCRIPTION
**Pull request by apollo13**
### Description
Added support to provide an initial token via the bootstrap HTTP API, similar to https://github.com/hashicorp/nomad/pull/12520

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [x] not a security concern

### CODE REVIEW

**Comment by jkirschner-hashicorp** [comment]
Hello @apollo13,

Thank you for submitting this PR for our consideration! To help with that consideration, can you provide more context on:
- The need that this PR is intended to address (since there is no linked Github issue)?
- The ways in which this PR addresses said need in a way that is more suitable for your use case than the existing Consul server agent configuration option [`acl.tokens.initial_management`](https://www.consul.io/docs/agent/config/config-files#acl_tokens_initial_management)?
  - As an aside, if you weren't previously aware of the `acl.tokens.initial_management` option, that suggests an opportunity for us to improve the ACL bootstrapping docs

**Comment by apollo13** [comment]
Hi @jkirschner-hashicorp,

> Thank you for submitting this PR for our consideration! To help with that consideration, can you provide more context on:

Sure, thank you for answering.

> * The need that this PR is intended to address (since there is no linked Github issue)?

I want to be able to programmatically bootstrap a Consul cluster. If nothing else, the hashicorp products should preferably behave the same way and this makes it nicely match nomad. 

> * The ways in which this PR addresses said need in a way that is more suitable for your use case than the existing Consul server agent configuration option [`acl.tokens.initial_management`](https://www.consul.io/docs/agent/config/config-files#acl_tokens_initial_management)?

I am aware of the initial management token but I rather not write (even if it is just an initial token) secrets like that onto the harddisk of a server if not needed. I understand that the raft/storage-layer will write the token anyways, but that storage layer is ecxpected to hold sensitive data (connect key material among others and in the worst case kv data is sensitive as well).

To make matters worse I am trying to precreate usable containers/vm images. The less key material I have to roll out the better. Initial server bootstrap works nicely via mdns cloud-join (got that working for nomad & consul). I understand that I will probably have to put other tokens into the config as well, but the fallout of those other tokens would be far less than the root key…

I hope this helps; I realize I am probably not making the best arguments for this feature -- it is not an absolute must-have (I understand that), but nevertheless it would make things easier for a few people out here.

**Comment by jkirschner-hashicorp** [comment]
Thank you for the context! I'll share this with the team.

**Comment by apollo13** [comment]
//Unstable

File: command/acl/bootstrap/bootstrap.go
Comment (rboyer) [inline command/acl/bootstrap/bootstrap.go:67]: There is already a [CLI helper](https://github.com/hashicorp/consul/blob/main/command/helpers/helpers.go#L57) to assist with this sequence of operations that we should use:

```suggestion
		terminalToken, err = helpers.LoadDataSourceNoRaw(args[0], os.Stdin)
```
Note that this is going to return a `string` already, so the typing for `terminalToken` needs fixing too. You could probably parse it directly into `var initialManagementToken string`.

File: command/acl/bootstrap/bootstrap.go
Comment (rboyer) [inline command/acl/bootstrap/bootstrap.go:75]: Since we know further down in the processing pipeline that we'll only accept a valid UUID for this field, it might be simpler to avoid any whitespace issues here by using `strings.TrimSpace` instead.

**Review by rboyer — CHANGES_REQUESTED** [review]
Great job on this! I have a variety of small comments and suggestions I'm adding below but none of them should be surprising or tricky to clean up before merging.

One of the longer comments is not actionable and is pointing out a `msgpack` non-issue to a future PR archaeologist.

File: agent/acl_endpoint.go
Comment (jkirschner-hashicorp) [reply agent/acl_endpoint.go:37]: Is this a line of code that a future dev might look at and go "wait, this looks like it might be wrong..."? If so, we might want an in-source comment. I defer to @rboyer

File: command/acl/bootstrap/bootstrap_test.go
Comment (kisunji) [inline command/acl/bootstrap/bootstrap_test.go:113]: Could we avoid using `ioutil`? Our linters will probably flag this as deprecated.
```suggestion
	f, err := os.CreateTemp("", "consul-token.token")
```

File: command/acl/bootstrap/bootstrap_test.go
Comment (kisunji) [inline command/acl/bootstrap/bootstrap_test.go:118]:
```suggestion
	err = os.WriteFile(f.Name(), []byte("2b778dd9-f5f1-6f29-b4b4-9a5fa948757a"), 0700)
```

File: command/acl/bootstrap/bootstrap_test.go
Comment (kisunji) [inline command/acl/bootstrap/bootstrap_test.go:5]:
```suggestion
```

**Comment by apollo13** [comment]
Thank you for the reviews. I will need a few days to get this updated.

File: agent/consul/acl_endpoint.go
Comment (apollo13) [reply agent/consul/acl_endpoint.go:223]: I `checkTokenUUID` also checks that the token does not start with `ACLReservedPrefix`, I'd like to at least keep that check. Should I switch to using `ACLIDReserved(secret)`?

File: agent/structs/acl.go
Comment (apollo13) [reply agent/structs/acl.go:1354]: I don't have strong feelings; I mirrored nomad: https://github.com/hashicorp/nomad/blob/fbe9f590489e291b3abcfb0fb1548a9f435a5e2d/nomad/structs/structs.go#L12169 -- I'd be fine with changing it if you think another name is better.

**Comment by apollo13** [comment]
@kisunji Any chance of another review?

**Review by rboyer — APPROVED** [review]

**Pull request merged**
